[FIXED JENKINS-31396] Add the submitterParameter property

jswager opened the pull request:
Add the submitterParameter property, used to record the submitter of an input step.

It will be treated like any other parameter on the input.  If the only value returned, will be returned directly. If other parameters, will be returned with the specified name.

Intended to fix https://issues.jenkins-ci.org/browse/JENKINS-31396

jswager commented:
@jglick - what is the process of getting a code review and merging the code if it's acceptable?

jglick commented on src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java:
```diff
@@ -301,8 +301,15 @@ private Object parseValue(StaplerRequest request) throws ServletException, IOExc
-        // TODO: perhaps we should return a different object to allow the workflow to look up
-        // who approved it, etc?
+        // If requested, add the approver to the list of returned parameters.
+        String submitter = input.getSubmitter();
```
No, this is just the definition of the `submitter` parameter in the script—includes no runtime information. Your test does not catch the mistake because it is hard-coding a single username, rather than using a comma-separated list or a `SecurityRealm` group name.

You meant to check `Jenkins.getAuthentication`; see elsewhere in this class.

jglick commented on src/test/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepTest.java:
```diff
@@ -165,6 +165,51 @@ public void test_submitters() throws Exception {
+
+        while (b.getAction(InputAction.class)==null) {
+            e.waitForSuspension();
+        }
```
Can all be done more simply with `j.waitForMessage`.

jglick commented on src/test/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepTest.java:
```diff
@@ -165,6 +165,51 @@ public void test_submitters() throws Exception {
+        System.out.println(b.getLog());
+        String v = b.getLog();
+        System.out.println(v);
+        assertTrue(b.getLog().contains("after: alice"));
```
Can all be replaced with

```java
j.assertLogContains("after: alice", b);
```

jswager commented:
Made documentation and testing changes based on review comments.  Also fixed the bug in my original logic.

Thank you for the review @jglick  - hopefully I got all the points you were flagging.

jswager commented:
All tests are passing now.  I updated my test that allowed the broken logic to pass such that it will catch such breaks in the future.  I hope.

jglick commented on src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java:
```diff
@@ -301,8 +301,16 @@ private Object parseValue(StaplerRequest request) throws ServletException, IOExc
-        // who approved it, etc?
+        // If requested, add the approver to the list of returned parameters.
+        String submitter = input.getSubmitter();
+        if (submitter != null) {
```
This means that if you write

```groovy
input message: 'x', submitterParameter: 'user'
```

you will silently get a null return value. That seems wrong to me—you may want to know who the submitter was, even if you are not restricting them in any particular way.

I think this `if`-block is just left over from the previous mistake, and you can simply ignore `submitter` here.

(Even if it *is* intentional to only allow `submitterParameter` when `submitter` is also set, this should be emphasized in HTML documentation, and checked in form validation in `InputStep.DescriptorImpl`.)

jglick commented on src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java:
```diff
@@ -301,8 +301,16 @@ private Object parseValue(StaplerRequest request) throws ServletException, IOExc
+        // If requested, add the approver to the list of returned parameters.
+        String submitter = input.getSubmitter();
+        if (submitter != null) {
+            Authentication a = Jenkins.getAuthentication();
```
Could be moved inside the `if`-block, since the value is used iff `valueName != null`.

jglick commented on src/test/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepTest.java:
```diff
@@ -165,6 +165,44 @@ public void test_submitters() throws Exception {
+        // job setup
+        WorkflowJob foo = j.jenkins.createProject(WorkflowJob.class, "foo");
+        foo.setDefinition(new CpsFlowDefinition(StringUtils.join(Arrays.asList(
+                "echo('before');",
```
This `echo` statement could be omitted I think.

jswager commented:
@jglick - Thank you for the suggestion in the review.  After thinking about the use case you describe (no submitter, but submitterParameter specified), I agree - that would be useful scenario to get the submitter regardless.  So I modified the code to match that.  And added another test to cover that particular method.

jglick commented on src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java:
```diff
@@ -301,8 +301,13 @@ private Object parseValue(StaplerRequest request) throws ServletException, IOExc
-        // who approved it, etc?
+        // If a destination value is specified, push the submitter to it.
+        String valueName = input.getSubmitterParameter();
+        if (valueName != null && !valueName.isEmpty()) {
```
Impossible for it to be empty, since you already called `fixEmptyAndTrim`.

forkbomber commented:
Thanks for contribution!
JFYI it seems that ok-type input return _null_ even when `submitterParameter` is passed as an argument (works fine for parameterized inputs though).

brleite commented:
@forkbomber 
The same happens with me. When I proceed through Stage View, the submitterParameter works. But when I proceed through Console Output, the input step returns a null object when no parameters are set.

jswager commented:
@forkbomber @brleite - could one of you please submit a JIRA incident with the details of this problem?  That way, it can be tracked in a way that other Jenkins users can see.  Thanks!
